### PR TITLE
Raise 100kB transaction size limit from Sapling activation

### LINKED ISSUE #2864: [Sapling] Raise the 100000-byte tx size limit
Before launch we imposed a consensus rule limiting the size of a transaction to 100000 bytes (Bitcoin already had this limit as a standard rule). #1475 #1501

A substantial part of the rationale for this limit was the quadratic hashing problem. Overwinter will fix that problem, so it isn't clear that 100000 bytes will still be the right limit at that point. We should consider raising it (which we can do straightforwardly because Overwinter is a bilateral hard fork).

### PR DESCRIPTION
**Pull request by str4d**
Closes #2864.

### CODE REVIEW

File: src/consensus/consensus.h [hunk 1]
```diff
@@ -23,7 +23,8 @@ static const unsigned int MAX_BLOCK_SIZE = 2000000;
 /** The maximum allowed number of signature check operations in a block (network rule) */
 static const unsigned int MAX_BLOCK_SIGOPS = 20000;
 /** The maximum size of a transaction (network rule) */
-static const unsigned int MAX_TX_SIZE = 100000;
+static const unsigned int MAX_TX_SIZE_BEFORE_SAPLING = 100000;
+static const unsigned int MAX_TX_SIZE = MAX_BLOCK_SIZE;
```
Comment (ebfull) [inline src/consensus/consensus.h]: `MAX_TX_SIZE` should be `MAX_TX_SIZE_AFTER_SAPLING` for explicitness and so that we have to touch every bit of code using the constant.

Comment (arielgabizon) [reply src/consensus/consensus.h]: Maybe, given that there will be more upgrades,  `MAX_TX_SIZE_SAPLING` is better without the AFTER ?

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -3482,6 +3482,7 @@ UniValue z_getoperationstatus_IMPL(const UniValue& params, bool fRemoveFinishedO
 // If input notes are small, we might actually require more than one joinsplit per zaddr output.
 // For now though, we assume we use one joinsplit per zaddr output (and the second output note is change).
 // We reduce the result by 1 to ensure there is room for non-joinsplit CTransaction data.
+#define Z_SENDMANY_MAX_ZADDR_OUTPUTS_BEFORE_SAPLING    ((MAX_TX_SIZE_BEFORE_SAPLING / GetSerializeSize(JSDescription(), SER_NETWORK, PROTOCOL_VERSION)) - 1)
 #define Z_SENDMANY_MAX_ZADDR_OUTPUTS    ((MAX_TX_SIZE / GetSerializeSize(JSDescription(), SER_NETWORK, PROTOCOL_VERSION)) - 1)
```
Comment (arielgabizon) [inline src/wallet/rpcwallet.cpp]: so if we agree with @ebfull 's comment above this should also have `AFTER_SAPLING` added?

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3482,7 +3482,8 @@ UniValue z_getoperationstatus_IMPL(const UniValue& params, bool fRemoveFinishedO
 // If input notes are small, we might actually require more than one joinsplit per zaddr output.
 // For now though, we assume we use one joinsplit per zaddr output (and the second output note is change).
 // We reduce the result by 1 to ensure there is room for non-joinsplit CTransaction data.
-#define Z_SENDMANY_MAX_ZADDR_OUTPUTS    ((MAX_TX_SIZE / GetSerializeSize(JSDescription(), SER_NETWORK, PROTOCOL_VERSION)) - 1)
+#define Z_SENDMANY_MAX_ZADDR_OUTPUTS_BEFORE_SAPLING    ((MAX_TX_SIZE_BEFORE_SAPLING / GetSerializeSize(JSDescription(), SER_NETWORK, PROTOCOL_VERSION)) - 1)
+#define Z_SENDMANY_MAX_ZADDR_OUTPUTS    ((MAX_TX_SIZE_AFTER_SAPLING / GetSerializeSize(JSDescription(), SER_NETWORK, PROTOCOL_VERSION)) - 1)
```
Comment (arielgabizon) [inline src/wallet/rpcwallet.cpp]: Should this also have a `AFTER_SAPLING` for clarity, consistency, and to make sure we're changing the code correctly in all places?

File: src/wallet/rpcwallet.cpp (hunk 2 above)
Comment (arielgabizon) [reply src/wallet/rpcwallet.cpp]: Still, it seems Sean's arguments that adding `AFTER_SAPLING` forces you to check that no bugs were introduced in all places is valid. And it's a small change. I wouldn't insist if it didn't relate to my second worry - the duplicate code between the three rpc methods. And I wonder if we can refactor some of it in this PR.

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (Eirik0) [reply src/wallet/rpcwallet.cpp]: Looked at the outdated comments. We do refer to this constant in an error message just below where it is defined. This will be incorrect until the upgrade, but perhaps this is not a big concern. Also, it is mentioned in the comment above that this may not be completely accurate in the case where there are multiple small notes.

On a side note, should Z_SENDMANY_MAX_ZADDR_OUTPUTS depend on the size of SpendDescritption and/or OutputDescription rather than JSDescription?

Comment (str4d) [reply src/wallet/rpcwallet.cpp]: Hmm, good point about SpendDescription and OutputDescription; they will need to be included in the sum (not replacing JSDescription), since we want z-addr outputs to be either. I'll go make adjustments (though most of them won't be relevant until we do #3215).

**Comment by str4d** [comment]
Added a test of the Sapling transaction size boundary.

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -3613,17 +3615,34 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
         nTotalOut += nAmount;
     }
 
-    // Check the number of zaddr outputs does not exceed the limit.
-    if (zaddrRecipients.size() > Z_SENDMANY_MAX_ZADDR_OUTPUTS)  {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, too many zaddr outputs");
+    int nextBlockHeight = chainActive.Height() + 1;
+    CMutableTransaction mtx;
+    mtx.fOverwintered = true;
+    mtx.nVersionGroupId = SAPLING_VERSION_GROUP_ID;
```
Comment (arielgabizon) [inline src/wallet/rpcwallet.cpp]: This determining of the version does not seem related to the transaction size limit.
Also wondering if it can be refactored to a generic method that does it according to the active upgrades

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -3633,8 +3652,8 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
         txsize += CTXOUT_REGULAR_SIZE;      // There will probably be taddr change
     }
     txsize += CTXOUT_REGULAR_SIZE * taddrRecipients.size();
-    if (txsize > MAX_TX_SIZE) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Too many outputs, size of raw transaction would be larger than limit of %d bytes", MAX_TX_SIZE ));
+    if (txsize > max_tx_size) {
```
Comment (arielgabizon) [inline src/wallet/rpcwallet.cpp]: Seems, the computation of `max_tx_size` in all three methods should be refactored somewhere, especially since it gets a little messy with looking at what are the active upgrades, which also distracts from what else is going on in the method.
Btw, I'm wondering why the transparent sending methods like `sendrawtransaction` do not require similar changes.

Comment (arielgabizon) [reply src/wallet/rpcwallet.cpp]: (I'm aware on argument for not refactoring is that some of this is temporary till upgrade after Sapling activation; still, I'm worried about the code getting entangled with these if's relating to what are the active upgrades)

File: src/wallet/rpcwallet.cpp (hunk 4 above)
Comment (str4d) [reply src/wallet/rpcwallet.cpp]: Yeah, `CreateNewContextualCMutableTransaction()` would likely end up being what we want to use. There are more places than this PR though, so we should do so in a separate PR to ensure that it makes sense everywhere.
